### ISSUE TITLE
Casting of a union type variable doesn't work if interface has all optional properties

### ISSUE THREAD

**Issue by piotrtomiak** [Microsoft/TypeScript#15289]
**TypeScript Version:**  at least 2.0.0 - 2.3.0

**Code**

```ts
interface ITest1 { IsTest: boolean; }
interface ITest2 { IsTest?: boolean; }

let case1: ITest1 | string;
case1 = case1 as string;
case1.indexOf("_"); // <-- no error

let case2: ITest2 | string;
case2 = case2 as string;
case2.indexOf("_"); // <-- ERROR
```

**Expected behavior:**
In the second case there should be no error reported, same as in the first one.  `case2` variable should have type `string` after casting `case2 = case2 as string`

**Actual behavior:**
When union of types contain interface with all optional properties than casting as one of the types does not change the type of a variable to more specific one.

**Comment by jesseschalken** [comment]
It seems when assigning to a variable, the type of the variable is only narrowed to members of the union that are compatible with the value assigned, and since `string` is compatible with `ITest2`, the type of `case2` stays as `ITest2 | string`.

Another example:

```ts
class A {
    aMethod() {}
}

class B extends A {
    bMethod() {}
}

class C {
    cMethod() {}
}

let x: A | B | C;
x = new C();
x.cMethod(); // ok
x = new B();
x.bMethod(); // error, bMethod() doesn't exist on A | B
```

**Comment by mhegazy** [comment]
This issue seems to have been already addressed or is unactionable at the moment. I am auto-closing it for now.

**Comment by piotrtomiak** [comment]
Behavior of TS seems to be wrong. Issue described by @jesseschalken confirms it. When TS knows exact type of the variable, it should narrow it to that exact type and not to all compatible types, as with more narrowed type I can call or access fields of the more generic types as well.

In @jesseschalken example if `x` would get type `B` instead of `A | B` I would still be able to call `aMethod`. Current behaviour is limiting and incorrect. I don't understand why this issue was closed.

**Comment by RyanCavanaugh** [comment]
This isn't a bug per se, but the behavior is pretty terrible.

We need a comprehensive fix that does something more sensible with degenerate union types.

**Comment by mhegazy** [comment]
@RyanCavanaugh, so what is the proposal here?

**Comment by RyanCavanaugh** [comment]
```ts
let x: A | B | C;
x = new C();
x.cMethod(); // ok

// Here, when we we narrow 'x' to the assignment source,
// we should eliminate any type from the produced union
// which is a supertype of all other constituents
x = new B();
x.bMethod();
```

**Comment by mhegazy** [comment]
but it is possible that `C` as a type is not in the type of `x`, so which ones you keep and which ones you narrow away? is it nominal? or is it structural?

**Comment by RyanCavanaugh** [comment]
I'm not sure what you mean.

This can all be done "correctly" with structural checks; no need for anything nominal.

**Comment by mhegazy** [comment]
```ts
type C = { a: number };
type NotJustC = { a: number, b?: string };

declare function getC(): C;

let x: number | NotJustC;
x = getC(); // is this C?
x.b = "b";  
```

**Comment by RyanCavanaugh** [comment]
It'd be `NotJustC`.

Given an assigning expression and a target union type, we have:
 * *Matching* constituents of the union type (`A` and `B`, above, or `NotJustC` below) where the expression type is assignable to that type
 * *Nonmatching* constituents (`C` / `number`)

Today we take all the *matching* constituents (`getAssignmentReducedType`) and make a union out of them, during which we perform subtype reduction.

What I'm proposing is that instead of just making a union type of the matching constituents, we throw away any matching constituent that is a supertype of another constituent. That would do nothing in your example, but remove `A` from the prior example.

**Comment by betopompolo** [comment]
@RyanCavanaugh @mhegazy any updates about this proposal?
